### PR TITLE
Add use_sim_time option to record verb

### LINKED ISSUE #764: Fill SerializedBagMessage.time_stamp with simulation time when recording a bag from a simulation
## Description

It seems that when recording (with` ros2 bag record` cli) a bag of a ros2 app running from simulation (and with a /clock topic), the SerializedBagMessage.time_stamp is set to the system time and not the simulation /clock time.
It would be convenient to fix in order not to run into issues of time reference when replaying a bag recorded from a simulation.
(or maybe it is already the case through an arg/config that I missed)

## Implementation Notes / Suggestions
I don't know what's the code path from the record cli, but if the underlying recording process is rclcpp::Node, I guess it is just a matter of switching its use_sim_time parameter to True

### PR DESCRIPTION
**Pull request by KeisukeShima**
Closes #764 

Based on @MichaelOrlov  's ([comment](https://github.com/ros2/rosbag2/issues/764#issuecomment-1085312999)), I have created a PR to reflect use_sim_time in the record verb.

## Test procedure

Before testing, please prepare a bag file.

* Terminal 1
```sh
ros2 bag play --clock 10 ORIGINAL_BAG
```

* Terminal 2
```sh
ros2 bag record --use-sim-time -a -o RECORDED_BAG
```

Then compare the values of `/clock`. They should be almost the same.
ex.

* Terminal 1
```sh
ros2 bag play --clock 10 [ORIGINAL_BAG|RECORDED_BAG]
```

* Terminal 2
```sh
ros2 topic echo /clock rosgraph_msgs/Clock > output
```

Signed-off-by: Keisuke Shima <19993104+KeisukeShima@users.noreply.github.com>

### CODE REVIEW

**Review by jhdcs — COMMENTED** [review]
Just a brief question on one part. This seems to be ok for the most part, though I feel that others who were more active in the discussion should take a look first before approval.

File: ros2bag/ros2bag/verb/record.py
Comment (KeisukeShima) [reply ros2bag/ros2bag/verb/record.py:162]: No default value needs to be specified here. According to the documentation, `store_true` will set the default value to False.

> 'store_true' and 'store_false' - These are special cases of 'store_const' used for storing the values True and False respectively. In addition, they create default values of False and True respectively.

https://docs.python.org/3/library/argparse.html#action

**Comment by KeisukeShima** [comment]
> Just a brief question on one part. This seems to be ok for the most part, though I feel that others who were more active in the discussion should take a look first before approval.

@jhdcs 
Yes, I agree. I have asked @MichaelOrlov  and @emersonknapp to review [here](https://github.com/ros2/rosbag2/issues/764#issuecomment-1109674124).
If possible, could you set them up as reviewers?

**Comment by MichaelOrlov** [comment]
@KeisukeShima Sorry for the delay. I will review tomorrow.

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (MichaelOrlov) [inline rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp:45]: We actually don't need to prepare vector of messages. We need to prepare only one message. Publication manager will repeat it anyway and will not operate with vector.

**Review by MichaelOrlov — CHANGES_REQUESTED** [review]
@KeisukeShima Thank you for your contribution.
Overall implementation looks good for me. Thank you for taking care about timestamp for bag split logic.

Although I have a few suggestions for improvements in unit tests.

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (MichaelOrlov) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp:61]: Ok. I've figured out the test works because we do initial discovery on start even if discovery disabled.

Nevertheless still highly recommend to provide list of topics explicitly to avoid interference with unexpected messages.

File: ros2bag/ros2bag/verb/record.py
Comment (KeisukeShima) [reply ros2bag/ros2bag/verb/record.py]: I see, so you set the argument explicitly for easy understanding. Understood.
d1f3abd

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (KeisukeShima) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: Thanks! I didn't know that I needed to add a topic list to RecorderOption.
4d37642

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (KeisukeShima) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: I see. I removed the verbose flag.
4d37642

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (KeisukeShima) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: Publication manager is very useful. I have modified it as suggested.
ecd0fa6

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (KeisukeShima) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: This is a very useful function, but I need to know the timestamp of the `string_messages`. And it is in `rosbag2_storage::SerializedBagMessage`, not in `test_msgs::msg::Strings`.
Should I create a new filter function in record_integration_fixture.hpp to get `SerializedBagMessage` to get the SerializedBagMessage?

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (wep21) [inline rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp:84]:
```suggestion
  EXPECT_THAT(string_messages.[0]->time_stamp, 1234567890123456789);
```
```suggestion
  EXPECT_THAT(string_messages[0]->time_stamp, 1234567890123456789);
```

Comment (wep21) [inline rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp:84]:
```suggestion
  EXPECT_THAT(string_messages.[0]->time_stamp, 1234567890123456789);
```
```suggestion
  EXPECT_THAT(string_messages[0]->time_stamp, 1234567890123456789);
```

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (MichaelOrlov) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: @KeisukeShima If you will not specify topic list it will record all available topics. Including `/rosout` which will interfere with tests expectations. 
Thanks for adding list of topics.

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (MichaelOrlov) [inline rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp:43]: By default shared pointer is used to be assigned to null pointer if no constructor specified.
Most probably it will be access by null address in next line and syshalt.  
It will be safer to specify constructor explicitly.
```cpp
auto clock_message = std::make_shared<rosgraph_msgs::msg::Clock>();
```

Comment (KeisukeShima) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: Thank you. I've fixed the code.
c77926d

File: rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp
Comment (KeisukeShima) [reply rosbag2_transport/test/rosbag2_transport/test_record_all_use_sim_time.cpp]: Thanks. I've fixed the code.
c77926d

**Review by MichaelOrlov — APPROVED** [review]
@KeisukeShima LGTM if CI will be running green.

**Comment by gbiggs** [comment]
Have some CI!

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/16685/)
- Linux-aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11272/)
- Windows [Build Status](https://ci.ros2.org/job/ci_windows/17086/)

**Comment by MichaelOrlov** [comment]
@gbiggs Thanks for running CI for this PR!

**Comment by KeisukeShima** [comment]
@MichaelOrlov Thanks for reviewing this PR and fixing the test #1014 
I would like to backport this PR to Humble before the next release. Do I need to create a new PR?
https://discourse.ros.org/t/humble-release-candidate-upcoming-ros-distro-freeze/25593

**Comment by MichaelOrlov** [comment]
@KeisukeShima Yes please create new PR for backporting which will be targeting `Humble` branch and add @audrow as one of the reviewer.

**Comment by MichaelOrlov** [comment]
@KeisukeShima BTW I am afraid that `Humble` branch already frozen for new features.  
@audrow Can you please clarify?

**Comment by KeisukeShima** [comment]
@MichaelOrlov Thanks for the advice. Even if it is not available in the next release, I think it is worth backporting. I will therefore create a PR.

**Pull request merged**
